core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java

```diff
@@ -0,0 +1,160 @@
+        InternalAggregations aggregations = aggregationsTests.createTestInstance();
+        Suggest suggest = SuggestTests.createTestItem();
+        SearchProfileShardResults profileShardResults = SearchProfileShardResultsTests.createTestItem();
+
+        InternalSearchResponse internalSearchResponse = new InternalSearchResponse(hits, aggregations, suggest, profileShardResults,
+                timedOut, terminatedEarly, numReducePhases);
+        return new SearchResponse(internalSearchResponse, null, totalShards, successfulShards, tookInMillis, failures);
+    }
+
+    public void testFromXContent() throws IOException {
+        // the "_shard/total/failures" section makes if impossible to directly compare xContent, so we omit it here
+        SearchResponse response = createTestItem(false);
+        XContentType xcontentType = randomFrom(XContentType.values());
+        boolean humanReadable = randomBoolean();
+        final ToXContent.Params params = new ToXContent.MapParams(singletonMap(RestSearchAction.TYPED_KEYS_PARAM, "true"));
+        BytesReference originalBytes = toShuffledXContent(response, xcontentType, params, humanReadable);
+        XContentParser parser = createParser(xcontentType.xContent(), originalBytes);
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+        SearchResponse parsed = SearchResponse.fromXContent(parser);
+        assertToXContentEquivalent(originalBytes, XContentHelper.toXContent(parsed, xcontentType, params, humanReadable), xcontentType);
+        assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+        assertNull(parser.nextToken());
+    }
+
+    /**
+     * The "_shard/total/failures" section makes if impossible to directly compare xContent, because
+     * the failures in the parsed SearchResponse are wrapped in an extra ElasticSearchException on the client side.
+     * Because of this, in this special test case we compare the "top level" fields for equality
+     * and the subsections xContent equivalence independently
+     */
+    public void testFromXContentWithFailures() throws IOException {
+        SearchResponse response = createTestItem(true);
+        XContentType xcontentType = randomFrom(XContentType.values());
+        final ToXContent.Params params = new ToXContent.MapParams(singletonMap(RestSearchAction.TYPED_KEYS_PARAM, "true"));
+        BytesReference originalBytes = toShuffledXContent(response, xcontentType, params, randomBoolean());
+        XContentParser parser = createParser(xcontentType.xContent(), originalBytes);
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+        SearchResponse parsed = SearchResponse.fromXContent(parser);
+        // check that we at least get the same number of shardFailures
+        assertEquals(response.getShardFailures().length, parsed.getShardFailures().length);
```

Comment:
We have some tests that create a failure instance and its expected instance, maybe we could have something similar here and use ElasticsearchExceptionTests.randomExceptions()?